File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteTaskRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Log;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.firebase.DatabaseWrapper;
import com.krystianwsul.checkme.firebase.json.InstanceJson;
import com.krystianwsul.checkme.firebase.json.ScheduleWrapper;
import com.krystianwsul.checkme.firebase.json.TaskJson;
import com.krystianwsul.checkme.utils.ScheduleKey;

import junit.framework.Assert;

import java.util.HashMap;
import java.util.Map;

public class RemoteTaskRecord extends RemoteRecord {
    public static final String TASKS = "tasks";

    @NonNull
    private final DomainFactory mDomainFactory;

    @NonNull
    private final String mId;

    @NonNull
    private final RemoteProjectRecord mRemoteProjectRecord;

    @NonNull
    private final TaskJson mTaskJson;

    @NonNull
    private final Map<ScheduleKey, RemoteInstanceRecord> mRemoteInstanceRecords = new HashMap<>();

    @NonNull
    public final Map<String, RemoteSingleScheduleRecord> mRemoteSingleScheduleRecords = new HashMap<>();

    @NonNull
    public final Map<String, RemoteDailyScheduleRecord> mRemoteDailyScheduleRecords = new HashMap<>();

    @NonNull
    public final Map<String, RemoteWeeklyScheduleRecord> mRemoteWeeklyScheduleRecords = new HashMap<>();

    @NonNull
    public final Map<String, RemoteMonthlyDayScheduleRecord> mRemoteMonthlyDayScheduleRecords = new HashMap<>();

    @NonNull
    public final Map<String, RemoteMonthlyWeekScheduleRecord> mRemoteMonthlyWeekScheduleRecords = new HashMap<>();

    RemoteTaskRecord(@NonNull DomainFactory domainFactory, @NonNull String id, @NonNull RemoteProjectRecord remoteProjectRecord, @NonNull TaskJson taskJson) {
        super(false);

        mDomainFactory = domainFactory;
        mId = id;
        mRemoteProjectRecord = remoteProjectRecord;
        mTaskJson = taskJson;

        initialize();
    }

    RemoteTaskRecord(@NonNull DomainFactory domainFactory, @NonNull RemoteProjectRecord remoteProjectRecord, @NonNull TaskJson taskJson) {
        super(true);

        mDomainFactory = domainFactory;
        mId = DatabaseWrapper.INSTANCE.getTaskRecordId(remoteProjectRecord.getId());
        mRemoteProjectRecord = remoteProjectRecord;
        mTaskJson = taskJson;

        initialize();
    }

    private void initialize() {
        for (Map.Entry<String, InstanceJson> entry : mTaskJson.getInstances().entrySet()) {
            String key = entry.getKey();
            Assert.assertTrue(!TextUtils.isEmpty(key));

            ScheduleKey scheduleKey = RemoteInstanceRecord.Companion.stringToScheduleKey(mDomainFactory, mRemoteProjectRecord.getId(), key);

            InstanceJson instanceJson = entry.getValue();
            Assert.assertTrue(instanceJson != null);

            RemoteInstanceRecord remoteInstanceRecord = new RemoteInstanceRecord(false, mDomainFactory, this, instanceJson, scheduleKey);

            mRemoteInstanceRecords.put(scheduleKey, remoteInstanceRecord);
        }

        for (Map.Entry<String, ScheduleWrapper> entry : mTaskJson.getSchedules().entrySet()) {
            ScheduleWrapper scheduleWrapper = entry.getValue();
            Assert.assertTrue(scheduleWrapper != null);

            String id = entry.getKey();
            Assert.assertTrue(!TextUtils.isEmpty(id));

            if (scheduleWrapper.singleScheduleJson != null) {
                Assert.assertTrue(scheduleWrapper.dailyScheduleJson == null);
                Assert.assertTrue(scheduleWrapper.weeklyScheduleJson == null);
                Assert.assertTrue(scheduleWrapper.monthlyDayScheduleJson == null);
                Assert.assertTrue(scheduleWrapper.monthlyWeekScheduleJson == null);

                mRemoteSingleScheduleRecords.put(id, new RemoteSingleScheduleRecord(id, this, scheduleWrapper));
            } else if (scheduleWrapper.dailyScheduleJson != null) {
                Assert.assertTrue(scheduleWrapper.weeklyScheduleJson == null);
                Assert.assertTrue(scheduleWrapper.monthlyDayScheduleJson == null);
                Assert.assertTrue(scheduleWrapper.monthlyWeekScheduleJson == null);

                mRemoteDailyScheduleRecords.put(id, new RemoteDailyScheduleRecord(id, this, scheduleWrapper));
            } else if (scheduleWrapper.weeklyScheduleJson != null) {
                Assert.assertTrue(scheduleWrapper.monthlyDayScheduleJson == null);
                Assert.assertTrue(scheduleWrapper.monthlyWeekScheduleJson == null);

                mRemoteWeeklyScheduleRecords.put(id, new RemoteWeeklyScheduleRecord(id, this, scheduleWrapper));
            } else if (scheduleWrapper.monthlyDayScheduleJson != null) {
                Assert.assertTrue(scheduleWrapper.monthlyWeekScheduleJson == null);

                mRemoteMonthlyDayScheduleRecords.put(id, new RemoteMonthlyDayScheduleRecord(id, this, scheduleWrapper));
            } else {
                Assert.assertTrue(scheduleWrapper.monthlyWeekScheduleJson != null);

                mRemoteMonthlyWeekScheduleRecords.put(id, new RemoteMonthlyWeekScheduleRecord(id, this, scheduleWrapper));
            }
        }
    }

    @NonNull
    @Override
    protected TaskJson getCreateObject() {
        if (!getCreate()) { // because of duplicate functionality when converting local task
            mTaskJson.setInstances(Stream.of(mRemoteInstanceRecords.entrySet()).collect(Collectors.toMap(entry -> RemoteInstanceRecord.Companion.scheduleKeyToString(mDomainFactory, mRemoteProjectRecord.getId(), entry.getKey()), entry -> entry.getValue().getCreateObject())));
        }

        Map<String, ScheduleWrapper> scheduleWrappers = new HashMap<>();

        for (RemoteSingleScheduleRecord remoteSingleScheduleRecord : mRemoteSingleScheduleRecords.values())
            scheduleWrappers.put(remoteSingleScheduleRecord.getId(), remoteSingleScheduleRecord.getCreateObject());

        for (RemoteDailyScheduleRecord remoteDailyScheduleRecord : mRemoteDailyScheduleRecords.values())
            scheduleWrappers.put(remoteDailyScheduleRecord.getId(), remoteDailyScheduleRecord.getCreateObject());

        for (RemoteWeeklyScheduleRecord remoteWeeklyScheduleRecord : mRemoteWeeklyScheduleRecords.values())
            scheduleWrappers.put(remoteWeeklyScheduleRecord.getId(), remoteWeeklyScheduleRecord.getCreateObject());

        for (RemoteMonthlyDayScheduleRecord remoteMonthlyDayScheduleRecord : mRemoteMonthlyDayScheduleRecords.values())
            scheduleWrappers.put(remoteMonthlyDayScheduleRecord.getId(), remoteMonthlyDayScheduleRecord.getCreateObject());

        for (RemoteMonthlyWeekScheduleRecord remoteMonthlyWeekScheduleRecord : mRemoteMonthlyWeekScheduleRecords.values())
            scheduleWrappers.put(remoteMonthlyWeekScheduleRecord.getId(), remoteMonthlyWeekScheduleRecord.getCreateObject());

        mTaskJson.setSchedules(scheduleWrappers);

        return mTaskJson;
    }

    @NonNull
    public String getId() {
        return mId;
    }

    @NonNull
    @Override
    protected String getKey() {
        return mRemoteProjectRecord.getKey() + "/" + RemoteProjectRecord.PROJECT_JSON + "/" + TASKS + "/" + mId;
    }

    @NonNull
    String getProjectId() {
        return mRemoteProjectRecord.getId();
    }

    @NonNull
    public String getName() {
        return mTaskJson.getName();
    }

    public long getStartTime() {
        return mTaskJson.getStartTime();
    }

    @Nullable
    public Long getEndTime() {
        return mTaskJson.getEndTime();
    }

    @Nullable
    public String getNote() {
        return mTaskJson.getNote();
    }

    @Nullable
    public Integer getOldestVisibleYear() {
        return mTaskJson.getOldestVisibleYear();
    }

    @Nullable
    public Integer getOldestVisibleMonth() {
        return mTaskJson.getOldestVisibleMonth();
    }

    @Nullable
    public Integer getOldestVisibleDay() {
        return mTaskJson.getOldestVisibleDay();
    }

    @NonNull
    public Map<ScheduleKey, RemoteInstanceRecord> getRemoteInstanceRecords() {
        return mRemoteInstanceRecords;
    }

    public void setEndTime(long endTime) {
        Assert.assertTrue(getEndTime() == null);

        mTaskJson.setEndTime(endTime);
        addValue(getKey() + "/endTime", endTime);
    }

    public void setOldestVisibleYear(int oldestVisibleYear) {
        if (getOldestVisibleYear() != null && getOldestVisibleYear().equals(oldestVisibleYear))
            return;

        mTaskJson.setOldestVisibleYear(oldestVisibleYear);
        addValue(getKey() + "/oldestVisibleYear", oldestVisibleYear);
    }

    public void setOldestVisibleMonth(int oldestVisibleMonth) {
        if (getOldestVisibleMonth() != null && getOldestVisibleMonth().equals(oldestVisibleMonth))
            return;

        mTaskJson.setOldestVisibleMonth(oldestVisibleMonth);
        addValue(getKey() + "/oldestVisibleMonth", oldestVisibleMonth);
    }

    public void setOldestVisibleDay(int oldestVisibleDay) {
        if (getOldestVisibleDay() != null && getOldestVisibleDay().equals(oldestVisibleDay))
            return;

        mTaskJson.setOldestVisibleDay(oldestVisibleDay);
        addValue(getKey() + "/oldestVisibleDay", oldestVisibleDay);
    }

    public void setName(@NonNull String name) {
        Assert.assertTrue(!TextUtils.isEmpty(name));

        if (getName().equals(name))
            return;

        mTaskJson.setName(name);
        addValue(getKey() + "/name", name);
    }

    public void setNote(@Nullable String note) {
        if (TextUtils.isEmpty(getNote())) {
            if (TextUtils.isEmpty(note))
                return;
        } else {
            if (getNote().equals(note))
                return;
        }

        mTaskJson.setNote(note);
        addValue(getKey() + "/note", note);
    }

    @Override
    public void getValues(@NonNull Map<String, Object> values) {
        Assert.assertTrue(!getDeleted());
        Assert.assertTrue(!getCreated());
        Assert.assertTrue(!getUpdated());

        if (getDelete()) {
            Log.e("asdf", "RemoteTaskRecord.getValues deleting " + this);

            Assert.assertTrue(!getCreate());
            Assert.assertTrue(getUpdate() != null);

            setDeleted(true);
            values.put(getKey(), null);
        } else if (getCreate()) {
            Log.e("asdf", "RemoteTaskRecord.getValues creating " + this);

            Assert.assertTrue(getUpdate() == null);

            setCreated(true);

            values.put(getKey(), getCreateObject());
        } else {
            Assert.assertTrue(getUpdate() != null);

            if (!getUpdate().isEmpty()) {
                Log.e("asdf", "RemoteTaskRecord.getValues updating " + this);

                setUpdated(true);
                values.putAll(getUpdate());
            }

            for (RemoteInstanceRecord remoteInstanceRecord : mRemoteInstanceRecords.values())
                remoteInstanceRecord.getValues(values);

            for (RemoteSingleScheduleRecord remoteSingleScheduleRecord : mRemoteSingleScheduleRecords.values())
                remoteSingleScheduleRecord.getValues(values);

            for (RemoteDailyScheduleRecord remoteDailyScheduleRecord : mRemoteDailyScheduleRecords.values())
                remoteDailyScheduleRecord.getValues(values);

            for (RemoteWeeklyScheduleRecord remoteWeeklyScheduleRecord : mRemoteWeeklyScheduleRecords.values())
                remoteWeeklyScheduleRecord.getValues(values);

            for (RemoteMonthlyDayScheduleRecord remoteMonthlyDayScheduleRecord : mRemoteMonthlyDayScheduleRecords.values())
                remoteMonthlyDayScheduleRecord.getValues(values);

            for (RemoteMonthlyWeekScheduleRecord remoteMonthlyWeekScheduleRecord : mRemoteMonthlyWeekScheduleRecords.values())
                remoteMonthlyWeekScheduleRecord.getValues(values);
        }
    }

    @NonNull
    public RemoteInstanceRecord newRemoteInstanceRecord(@NonNull DomainFactory domainFactory, @NonNull InstanceJson instanceJson, @NonNull ScheduleKey scheduleKey) {
        RemoteInstanceRecord remoteInstanceRecord = new RemoteInstanceRecord(true, domainFactory, this, instanceJson, scheduleKey);
        Assert.assertTrue(!mRemoteInstanceRecords.containsKey(remoteInstanceRecord.getScheduleKey()));

        mRemoteInstanceRecords.put(remoteInstanceRecord.getScheduleKey(), remoteInstanceRecord);
        return remoteInstanceRecord;
    }

    @NonNull
    public RemoteSingleScheduleRecord newRemoteSingleScheduleRecord(@NonNull ScheduleWrapper scheduleWrapper) {
        RemoteSingleScheduleRecord remoteSingleScheduleRecord = new RemoteSingleScheduleRecord(this, scheduleWrapper);
        Assert.assertTrue(!mRemoteSingleScheduleRecords.containsKey(remoteSingleScheduleRecord.getId()));

        mRemoteSingleScheduleRecords.put(remoteSingleScheduleRecord.getId(), remoteSingleScheduleRecord);
        return remoteSingleScheduleRecord;
    }

    @NonNull
    public RemoteWeeklyScheduleRecord newRemoteWeeklyScheduleRecord(@NonNull ScheduleWrapper scheduleWrapper) {
        RemoteWeeklyScheduleRecord remoteWeeklyScheduleRecord = new RemoteWeeklyScheduleRecord(this, scheduleWrapper);
        Assert.assertTrue(!mRemoteWeeklyScheduleRecords.containsKey(remoteWeeklyScheduleRecord.getId()));

        mRemoteWeeklyScheduleRecords.put(remoteWeeklyScheduleRecord.getId(), remoteWeeklyScheduleRecord);
        return remoteWeeklyScheduleRecord;
    }

    @NonNull
    public RemoteMonthlyDayScheduleRecord newRemoteMonthlyDayScheduleRecord(@NonNull ScheduleWrapper scheduleWrapper) {
        RemoteMonthlyDayScheduleRecord remoteMonthlyDayScheduleRecord = new RemoteMonthlyDayScheduleRecord(this, scheduleWrapper);
        Assert.assertTrue(!mRemoteMonthlyDayScheduleRecords.containsKey(remoteMonthlyDayScheduleRecord.getId()));

        mRemoteMonthlyDayScheduleRecords.put(remoteMonthlyDayScheduleRecord.getId(), remoteMonthlyDayScheduleRecord);
        return remoteMonthlyDayScheduleRecord;
    }

    @NonNull
    public RemoteMonthlyWeekScheduleRecord newRemoteMonthlyWeekScheduleRecord(@NonNull ScheduleWrapper scheduleWrapper) {
        RemoteMonthlyWeekScheduleRecord remoteMonthlyWeekScheduleRecord = new RemoteMonthlyWeekScheduleRecord(this, scheduleWrapper);
        Assert.assertTrue(!mRemoteMonthlyWeekScheduleRecords.containsKey(remoteMonthlyWeekScheduleRecord.getId()));

        mRemoteMonthlyWeekScheduleRecords.put(remoteMonthlyWeekScheduleRecord.getId(), remoteMonthlyWeekScheduleRecord);
        return remoteMonthlyWeekScheduleRecord;
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteTaskRecord.kt
package com.krystianwsul.checkme.firebase.records

import android.text.TextUtils
import com.krystianwsul.checkme.MyCrashlytics
import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.firebase.RemoteProject
import com.krystianwsul.checkme.firebase.RemoteTask
import com.krystianwsul.checkme.firebase.json.InstanceJson
import com.krystianwsul.checkme.firebase.json.OldestVisibleJson
import com.krystianwsul.checkme.firebase.json.ScheduleWrapper
import com.krystianwsul.checkme.firebase.json.TaskJson
import com.krystianwsul.checkme.utils.RemoteCustomTimeId
import com.krystianwsul.checkme.utils.ScheduleKey
import com.krystianwsul.checkme.utils.time.Date

class RemoteTaskRecord<T : RemoteCustomTimeId> private constructor(
        create: Boolean,
        val domainFactory: DomainFactory,
        val id: String,
        private val remoteProjectRecord: RemoteProjectRecord<T>,
        private val taskJson: TaskJson) : RemoteRecord(create) {

    companion object {

        const val TASKS = "tasks"
    }

    val remoteInstanceRecords = mutableMapOf<ScheduleKey, RemoteInstanceRecord<T>>()

    val remoteSingleScheduleRecords: MutableMap<String, RemoteSingleScheduleRecord<T>> = HashMap()

    val remoteDailyScheduleRecords: MutableMap<String, RemoteDailyScheduleRecord<T>> = HashMap()

    val remoteWeeklyScheduleRecords: MutableMap<String, RemoteWeeklyScheduleRecord<T>> = HashMap()

    val remoteMonthlyDayScheduleRecords: MutableMap<String, RemoteMonthlyDayScheduleRecord<T>> = HashMap()

    val remoteMonthlyWeekScheduleRecords: MutableMap<String, RemoteMonthlyWeekScheduleRecord<T>> = HashMap()

    override val createObject: TaskJson
        // because of duplicate functionality when converting local task
        get() {
            if (update != null)
                taskJson.instances = remoteInstanceRecords.entries
                        .associateBy({ RemoteInstanceRecord.scheduleKeyToString(it.key) }, { it.value.createObject })
                        .toMutableMap()

            val scheduleWrappers = HashMap<String, ScheduleWrapper>()

            for (remoteSingleScheduleRecord in remoteSingleScheduleRecords.values)
                scheduleWrappers[remoteSingleScheduleRecord.id] = remoteSingleScheduleRecord.createObject

            for (remoteDailyScheduleRecord in remoteDailyScheduleRecords.values)
                scheduleWrappers[remoteDailyScheduleRecord.id] = remoteDailyScheduleRecord.createObject

            for (remoteWeeklyScheduleRecord in remoteWeeklyScheduleRecords.values)
                scheduleWrappers[remoteWeeklyScheduleRecord.id] = remoteWeeklyScheduleRecord.createObject

            for (remoteMonthlyDayScheduleRecord in remoteMonthlyDayScheduleRecords.values)
                scheduleWrappers[remoteMonthlyDayScheduleRecord.id] = remoteMonthlyDayScheduleRecord.createObject

            for (remoteMonthlyWeekScheduleRecord in remoteMonthlyWeekScheduleRecords.values)
                scheduleWrappers[remoteMonthlyWeekScheduleRecord.id] = remoteMonthlyWeekScheduleRecord.createObject

            taskJson.schedules = scheduleWrappers

            return taskJson
        }

    override val key get() = remoteProjectRecord.childKey + "/" + TASKS + "/" + id

    val projectId get() = remoteProjectRecord.id

    var name
        get() = taskJson.name
        set(name) {
            check(!TextUtils.isEmpty(name))

            if (name == taskJson.name)
                return

            taskJson.name = name
            addValue("$key/name", name)
        }

    val startTime get() = taskJson.startTime

    var endTime
        get() = taskJson.endTime
        set(value) {
            if (value == taskJson.endTime)
                return

            taskJson.endTime = value
            addValue("$key/endTime", value)
        }

    var note
        get() = taskJson.note
        set(note) {
            if (note == taskJson.note)
                return

            taskJson.note = note
            addValue("$key/note", note)
        }

    val oldestVisible
        get() = taskJson.oldestVisible
                .values
                .map { it.toDate() }
                .toMutableList()
                .apply {
                    if (taskJson.oldestVisibleYear != null && taskJson.oldestVisibleMonth != null && taskJson.oldestVisibleDay != null) {
                        add(Date(taskJson.oldestVisibleYear!!, taskJson.oldestVisibleMonth!!, taskJson.oldestVisibleDay!!))
                    } else {
                        if (taskJson.oldestVisibleYear != null || taskJson.oldestVisibleMonth != null || taskJson.oldestVisibleDay != null)
                            MyCrashlytics.logException(RemoteTask.MissingDayException("projectId: $projectId, taskId: $id, oldestVisibleYear: ${taskJson.oldestVisibleYear}, oldestVisibleMonth: ${taskJson.oldestVisibleMonth}, oldestVisibleDay: ${taskJson.oldestVisibleDay}"))
                    }
                }
                .min()

    constructor(domainFactory: DomainFactory, id: String, remoteProjectRecord: RemoteProjectRecord<T>, taskJson: TaskJson) : this(
            false,
            domainFactory,
            id,
            remoteProjectRecord,
            taskJson)

    constructor(domainFactory: DomainFactory, remoteProjectRecord: RemoteProjectRecord<T>, taskJson: TaskJson) : this(
            true,
            domainFactory,
            remoteProjectRecord.getTaskRecordId(),
            remoteProjectRecord,
            taskJson)

    init {
        if (taskJson.name.isEmpty())
            MyCrashlytics.logException(MissingNameException("taskKey: $key"))

        for ((key, instanceJson) in taskJson.instances) {
            check(!TextUtils.isEmpty(key))

            val (scheduleKey, remoteCustomTimeId) = RemoteInstanceRecord.stringToScheduleKey(remoteProjectRecord, key)

            val remoteInstanceRecord = RemoteInstanceRecord(
                    create,
                    this,
                    instanceJson,
                    scheduleKey,
                    key,
                    remoteCustomTimeId)

            remoteInstanceRecords[scheduleKey] = remoteInstanceRecord
        }

        for ((id, scheduleWrapper) in taskJson.schedules) {
            check(!TextUtils.isEmpty(id))

            when {
                scheduleWrapper.singleScheduleJson != null -> {
                    check(scheduleWrapper.dailyScheduleJson == null)
                    check(scheduleWrapper.weeklyScheduleJson == null)
                    check(scheduleWrapper.monthlyDayScheduleJson == null)
                    check(scheduleWrapper.monthlyWeekScheduleJson == null)

                    remoteSingleScheduleRecords[id] = RemoteSingleScheduleRecord(id, this, scheduleWrapper)
                }
                scheduleWrapper.dailyScheduleJson != null -> {
                    check(scheduleWrapper.weeklyScheduleJson == null)
                    check(scheduleWrapper.monthlyDayScheduleJson == null)
                    check(scheduleWrapper.monthlyWeekScheduleJson == null)

                    remoteDailyScheduleRecords[id] = RemoteDailyScheduleRecord(id, this, scheduleWrapper)
                }
                scheduleWrapper.weeklyScheduleJson != null -> {
                    check(scheduleWrapper.monthlyDayScheduleJson == null)
                    check(scheduleWrapper.monthlyWeekScheduleJson == null)

                    remoteWeeklyScheduleRecords[id] = RemoteWeeklyScheduleRecord(id, this, scheduleWrapper)
                }
                scheduleWrapper.monthlyDayScheduleJson != null -> {
                    check(scheduleWrapper.monthlyWeekScheduleJson == null)

                    remoteMonthlyDayScheduleRecords[id] = RemoteMonthlyDayScheduleRecord(id, this, scheduleWrapper)
                }
                else -> {
                    check(scheduleWrapper.monthlyWeekScheduleJson != null)

                    remoteMonthlyWeekScheduleRecords[id] = RemoteMonthlyWeekScheduleRecord(id, this, scheduleWrapper)
                }
            }
        }
    }

    private val uuid by lazy { domainFactory.uuid }

    private val oldestVisibleJson get() = taskJson.oldestVisible[uuid]

    override val children
        get() = remoteInstanceRecords.values +
                remoteSingleScheduleRecords.values +
                remoteDailyScheduleRecords.values +
                remoteWeeklyScheduleRecords.values +
                remoteMonthlyDayScheduleRecords.values +
                remoteMonthlyWeekScheduleRecords.values

    fun setOldestVisible(newOldestVisibleJson: OldestVisibleJson) {
        val oldOldestVisibleJson = oldestVisibleJson

        taskJson.oldestVisible[uuid] = newOldestVisibleJson

        if (oldOldestVisibleJson?.date != newOldestVisibleJson.date)
            addValue("$key/oldestVisible/$uuid/date", newOldestVisibleJson.date)

        if (oldOldestVisibleJson?.year != newOldestVisibleJson.year)
            addValue("$key/oldestVisible/$uuid/year", newOldestVisibleJson.year)

        if (oldOldestVisibleJson?.month != newOldestVisibleJson.month)
            addValue("$key/oldestVisible/$uuid/month", newOldestVisibleJson.month)

        if (oldOldestVisibleJson?.day != newOldestVisibleJson.day)
            addValue("$key/oldestVisible/$uuid/day", newOldestVisibleJson.day)

        if (newOldestVisibleJson.year != taskJson.oldestVisibleYear) {
            taskJson.oldestVisibleYear = newOldestVisibleJson.year
            addValue("$key/oldestVisibleYear", newOldestVisibleJson.year)
        }

        if (newOldestVisibleJson.month != taskJson.oldestVisibleMonth) {
            taskJson.oldestVisibleMonth = newOldestVisibleJson.month
            addValue("$key/oldestVisibleMonth", newOldestVisibleJson.month)
        }

        if (newOldestVisibleJson.day != taskJson.oldestVisibleDay) {
            taskJson.oldestVisibleDay = newOldestVisibleJson.day
            addValue("$key/oldestVisibleDay", newOldestVisibleJson.day)
        }
    }

    fun newRemoteInstanceRecord(remoteProject: RemoteProject<T>, instanceJson: InstanceJson, scheduleKey: ScheduleKey): RemoteInstanceRecord<T> {
        val remoteCustomTimeId = scheduleKey.scheduleTimePair
                .customTimeKey
                ?.let { remoteProject.getRemoteCustomTime(it.remoteCustomTimeId).id }

        val firebaseKey = RemoteInstanceRecord.scheduleKeyToString(scheduleKey)

        val remoteInstanceRecord = RemoteInstanceRecord(true, this, instanceJson, scheduleKey, firebaseKey, remoteCustomTimeId)
        check(!remoteInstanceRecords.containsKey(remoteInstanceRecord.scheduleKey))

        remoteInstanceRecords[remoteInstanceRecord.scheduleKey] = remoteInstanceRecord
        return remoteInstanceRecord
    }

    fun newRemoteSingleScheduleRecord(scheduleWrapper: ScheduleWrapper): RemoteSingleScheduleRecord<T> {
        val remoteSingleScheduleRecord = RemoteSingleScheduleRecord(this, scheduleWrapper)
        check(!remoteSingleScheduleRecords.containsKey(remoteSingleScheduleRecord.id))

        remoteSingleScheduleRecords[remoteSingleScheduleRecord.id] = remoteSingleScheduleRecord
        return remoteSingleScheduleRecord
    }

    fun newRemoteWeeklyScheduleRecord(scheduleWrapper: ScheduleWrapper): RemoteWeeklyScheduleRecord<T> {
        val remoteWeeklyScheduleRecord = RemoteWeeklyScheduleRecord(this, scheduleWrapper)
        check(!remoteWeeklyScheduleRecords.containsKey(remoteWeeklyScheduleRecord.id))

        remoteWeeklyScheduleRecords[remoteWeeklyScheduleRecord.id] = remoteWeeklyScheduleRecord
        return remoteWeeklyScheduleRecord
    }

    fun newRemoteMonthlyDayScheduleRecord(scheduleWrapper: ScheduleWrapper): RemoteMonthlyDayScheduleRecord<T> {
        val remoteMonthlyDayScheduleRecord = RemoteMonthlyDayScheduleRecord(this, scheduleWrapper)
        check(!remoteMonthlyDayScheduleRecords.containsKey(remoteMonthlyDayScheduleRecord.id))

        remoteMonthlyDayScheduleRecords[remoteMonthlyDayScheduleRecord.id] = remoteMonthlyDayScheduleRecord
        return remoteMonthlyDayScheduleRecord
    }

    fun newRemoteMonthlyWeekScheduleRecord(scheduleWrapper: ScheduleWrapper): RemoteMonthlyWeekScheduleRecord<T> {
        val remoteMonthlyWeekScheduleRecord = RemoteMonthlyWeekScheduleRecord(this, scheduleWrapper)
        check(!remoteMonthlyWeekScheduleRecords.containsKey(remoteMonthlyWeekScheduleRecord.id))

        remoteMonthlyWeekScheduleRecords[remoteMonthlyWeekScheduleRecord.id] = remoteMonthlyWeekScheduleRecord
        return remoteMonthlyWeekScheduleRecord
    }

    override fun deleteFromParent() = check(remoteProjectRecord.remoteTaskRecords.remove(id) == this)

    fun getScheduleRecordId() = remoteProjectRecord.getScheduleRecordId(id)

    fun getRemoteCustomTimeId(id: String) = remoteProjectRecord.getRemoteCustomTimeId(id)

    private class MissingNameException(message: String) : Exception(message)
}
